### PR TITLE
Fix #2114: Fix CharBuffer.wrap(CharSequence,Int,Int)

### LINKED ISSUE #2114: CharBufferWrappingACharSequenceTest is failing on JVM.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Note that `CharBuffer.wrap(CharSequence,Int,Int)` delegates to `StringCharBuffer.wrap` which is the source of the bug.

**Comment by nicolasstucki** [comment]
Review by @sjrd

File: javalib/src/main/scala/java/nio/StringCharBuffer.scala [hunk 1]
```diff
@@ -87,11 +87,10 @@ private[nio] final class StringCharBuffer private (
 
 private[nio] object StringCharBuffer {
   private[nio] def wrap(csq: CharSequence, csqOffset: Int, capacity: Int,
-      initialPosition: Int, initialLength: Int): CharBuffer = {
+      initialPosition: Int, initialLimit: Int): CharBuffer = {
     if (csqOffset < 0 || capacity < 0 || csqOffset+capacity > csq.length)
       throw new IndexOutOfBoundsException
-    val initialLimit = initialPosition + initialLength
-    if (initialPosition < 0 || initialLength < 0 || initialLimit > capacity)
+    if (initialPosition < 0 || initialLimit < 0 || initialLimit > capacity)
```
Comment (sjrd) [inline javalib/src/main/scala/java/nio/StringCharBuffer.scala]: Shouldn't the test in middle be `initialLimit < initialPosition`?

Comment (nicolasstucki) [inline javalib/src/main/scala/java/nio/StringCharBuffer.scala]: Yes

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala [hunk 2]
```diff
@@ -33,11 +33,14 @@ abstract class BaseBufferTest {
     assertEquals(0, allocBuffer(0).capacity)
 
     expectThrows(classOf[Exception], allocBuffer(-1))
+    expectThrows(classOf[Throwable], allocBuffer(0, -1, 1))
+    expectThrows(classOf[Throwable], allocBuffer(1, 0, 1))
+    expectThrows(classOf[Throwable], allocBuffer(0, 1, 0))
+    expectThrows(classOf[Throwable], allocBuffer(1, 0, 0))
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: `IndexOutOfBoundsException`?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala [hunk 3]
```diff
@@ -298,7 +287,7 @@ abstract class BaseBufferTest {
       assertEquals(elemFromInt(0), buf.get(0))
 
       buf.position(8)
-      if (!executingInJVM)
+      if (!executingInJVM) // throws BufferOverflowException on JVM
         expectThrows(classOf[ReadOnlyBufferException], buf.put(Array[ElementType](6, 7, 12)))
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: I guess we should also fix that, shouldn't we? That we be another issue.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala [hunk 4]
```diff
@@ -312,7 +301,7 @@ abstract class BaseBufferTest {
       buf.mark()
 
       if (!executingInJVM) {
-        // issues #2114: CharBufferWrappingACharSequenceTest
+        // throws IllegalArgumentException on JVM when executing compact()
         buf.compact()
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: Is that another issue?

**Comment by sjrd** [comment]
That's all.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala (hunk 2 above)
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: `UBE` in full opt :(

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala (hunk 3 above)
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: Yes, that would be another issue.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala (hunk 2 above)
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: How so?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala (hunk 3 above)
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: Added as #2190.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala (hunk 4 above)
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: Yes, it seems that I mislabeled it in the first place.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala (hunk 2 above)
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: I believe that it was failing with `UBE` with the wrapped typed arrays.

Furthermore the best I could do is `RuntimeException` as on the JVM the exception can be either 
`IndexOutOfBoundsException` or `IllegalArgumentException`.

Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: Ah right. OK

**Comment by nicolasstucki** [comment]
Then I guess that all comments were addressed without modification on the code.

**Comment by sjrd** [comment]
Shouldn't there be references to the GitHub issues in the comments about the remaining `!executingInJVM`?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala (hunk 4 above)
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/niobuffer/BaseBufferTest.scala]: Added as #2192.

**Comment by nicolasstucki** [comment]
Yes, just forgot to click on the button. Now it's there.

**Comment by sjrd** [comment]
LGTM

**Pull request merged**